### PR TITLE
pkgconfig: Generate -uninstalled.pc files

### PR DESCRIPTION
**Pull request by xclaesse**
Closes: #3472.

### CODE REVIEW

File: docs/markdown/Pkgconfig-module.md [hunk 1]
```diff
@@ -51,10 +51,19 @@ keyword arguments.
...
 
 Since 0.46 a `StaticLibrary` or `SharedLibrary` object can optionally be passed
 as first positional argument. If one is provided a default value will be
 provided for all required fields of the pc file:
 - `install_dir` is set to `pkgconfig` folder in the same location than the provided library.
 - `description` is set to the project's name followed by the library's name.
 - `name` is set to the library's name.
+
+Since 0.49.0 uninstalled pkg-config files are generated as well. They are
+located in `<build dir>/pkgconfig-uninstalled/*-uninstalled.pc`. It is useful to
```
Comment (jon-turney) [inline docs/markdown/Pkgconfig-module.md]: 'located in <filename>': either 'located in <directory>', or say 'called <filename>'.

'useful '-> 'sometimes useful'

File: docs/markdown/Pkgconfig-module.md [hunk 2]
```diff
@@ -51,10 +51,19 @@ keyword arguments.
...
+build projects against libraries built by meson without having to install them
+into a prefix. In order to do so, just set `PKG_CONFIG_PATH=<builddir>/pkgconfig-uninstalled`
+before building your application. That will cause pkg-config to use those pc files
+in priority and find libraries and headers from the meson builddir.
```
Comment (jon-turney) [inline docs/markdown/Pkgconfig-module.md]: 'pkg-config will prefer the `-uninstalled.pc` files, for compilation/linking against...'

File: docs/markdown/Pkgconfig-module.md [hunk 3]
```diff
@@ -51,10 +51,19 @@ keyword arguments.
 - `version` a string describing the version of this library
 - `d_module_versions` a list of module version flags used when compiling
    D sources referred to by this pkg-config file
+- `uninstalled_variables` used instead of the `variables` keyword argument, when
+  generating the uninstalled pkg-config file. Since *0.49.0*
```
[the remark this answers is not in the document]
Comment (xclaesse) [reply docs/markdown/Pkgconfig-module.md]: I wasn't sure about this, variables often contains paths that needs to be pointing to builddir and that's not something meson can do automatically. I think having them completely separate makes it clearer. For example if you have `variables=['a=1', 'b=2'], uninstalled_variables=['b=3']`it's not immediately clear if you have `a=1` in uninstalled or not.

I don't know...

**Comment by xclaesse** [comment]
test case added.

**Comment by xclaesse** [comment]
I'm not sure to understand the CI failure, with cygwin the library in builddir is named `libfoo-1.dll` and that doesn't match `-lfoo` ? It gets renamed on install?

File: docs/markdown/Pkgconfig-module.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jon-turney) [reply docs/markdown/Pkgconfig-module.md]: Yeah.  I guess there are no good solutions here.  But it means that non-path variables in uninstalled.pc files created by existing meson.build files will be wrong until they are explicitly added...

**Comment by jon-turney** [comment]
> I'm not sure to understand the CI failure, with cygwin the library in builddir is named `libfoo-1.dll` and that doesn't match `-lfoo` ? It gets renamed on install?

No, that would be bad.  

The thing the linker should be looking for is the import library libfoo.dll.a, but it seems that's not being named correctly.  I think that might be a bug with `name_prefix : ''` on cygwin.

**Comment by jon-turney** [comment]
Hmmm.

2d5172b51d9077030aa4d6432bb49ee32a1597f8 and 071264c9cb36b9c08940c5d06f48488400d594da makes this test pass for me.

But I'd like to think a bit more about how `name_prefix` should interact with import library naming.

**Comment by jon-turney** [comment]
> But I'd like to think a bit more about how `name_prefix` should interact with import library naming.

Naming the import library correctly in this way is only important when we install the library and subsequently link with it using -l, internal uses are fine as we just use the filename.

The relevant PR seems to be #899.  The pkgconfig module only handles 2 scenarios: (i) `name_prefix` is the default, or (ii) `name_prefix` is empty, and the library name starts with `lib`.  Anything else causes a warning, as generating linker flags is problematic.

Anyhow, please squash 071264c9cb36b9c08940c5d06f48488400d594da as a fix to your new testcase, and  add 2e62422cda2a9e380359b6a2318b51d622f0e185 as a fix to an existing bug.

**Comment by xclaesse** [comment]
@jon-turney I included both your patches in my PR. I simplified  https://github.com/mesonbuild/meson/commit/071264c9cb36b9c08940c5d06f48488400d594da to not save/restore env. I think you should still create its own PR for https://github.com/mesonbuild/meson/commit/2e62422cda2a9e380359b6a2318b51d622f0e185 because that looks like an important fix that shouldn't block on getting this -uninstalled pc files feature approved.

**Comment by xclaesse** [comment]
@jpakkane @nirbheek Would be great to have a review, I have patches blocked by this for GStreamer.

**Comment by jpakkane** [comment]
I really, really don't like the concept of -uninstalled. It gives the indication that we are committing to keeping some sort of stability for things inside the build dir, which we are never going to do. What is the actual use case for this? Who is using it? Could we solve the same problem in some other way instead?

**Comment by xclaesse** [comment]
We don't need to guarantee any stability here, because if we change the builddir structure, we can always update the -uninstalled generator accordingly.

We have gst-build that is a meson project that builds all gstreamer modules and deps as subprojects. We have a script `gst-uninstalled` that start a dev env with PKG_CONFIG_PATH, LD_LIBRARY_PATH, etc to be able to quickly build and run an app using the gstreamer version built by meson. That avoids having to run "ninja install" all the time.

So that's really a developer tool, not something distro will use of course.

File: mesonbuild/modules/pkgconfig.py [hunk 4]
```diff
@@ -276,16 +287,15 @@ def generate_libs_flags(libs):
                         install_dir = l.get_custom_install_dir()[0]
                         if install_dir is False:
                             continue
+                        if uninstalled:
+                            path = os.path.dirname(state.backend.get_target_filename_abs(l))
+                            Lflag = '-L%s' % self._escape(path)
```
Comment (textshell) [inline mesonbuild/modules/pkgconfig.py]: Can we get away with just using absolute paths to the library in the uninstalled case? I think in the c and c++ case it should really work and avoid polluting the link command line with directories that might contain internal files that would not have been installed.

File: mesonbuild/modules/pkgconfig.py [hunk 5]
```diff
@@ -297,22 +307,42 @@ def generate_libs_flags(libs):
                         if 'cs' not in l.compilers:
                             yield '-l%s' % lname
 
+            def generate_include_flags(libs):
```
Comment (textshell) [inline mesonbuild/modules/pkgconfig.py]: this name is confusing. I think it should be `generate_uninstalled_include_flags`.
Also it seems to do something quite different than the installed case.

The paths used for compilation of a library don't need to match what the library user expects. For one they might include tons of private things. And there are quite some libraries that get compiled with paths pointing inside the library specific include directory instead of pointing only to the directory containing it.

Of course the includes in the uninstalled case are a hard to get in the general case. I think the only really supportable case is when subdirs points to directories that are 1:1 (subdir matching the last elements of the path of all files) installed and where nothing else is placed into by other install related calls. I'm not sure how generated header fit into this, but i think there should be a way to allow them too.

**Comment by textshell** [comment]
I think it's important to have a way to use meson build libraries without installing them in many situations.

Installation into a common prefix gets very messy fast, because things are not really independent.

And as soon as a user has many independent leaf projects depending on a forest of libraries that are all unter active development there needs to be a reliable solution to manage that without having to rebuild everything as subprojects for each leaf project.

I've even made a wrapper to allow usage of uninstalled libraries for some of my projects. I don't think forcing installation to a temporary location is a good alternative either, because then you loose installation as the process that places the files in their final location. Often mixing development with uninstalled pieces of code and installation after everything is in a good state without recompiling the whole project is a desirable path.

That said i think we need to come up with a good solution for the include paths in the uninstalled variant. I believe we might have to some some restrictions on possible directory structures. I.e. if the includes are restructured on install i think we should not try to support that. But if a project wants to have a good uninstalled story and places all includes already in suitable directories we should support that.

File: test cases/common/48 pkgconfig-gen/meson.build [hunk 6]
```diff
@@ -37,7 +36,8 @@ test('pkgconfig-validation', pkgconfig,
 # Test that name_prefix='' and name='libfoo' results in '-lfoo'
 lib2 = shared_library('libfoo', 'simple.c',
   name_prefix : '',
-  version : libver)
+  version : libver,
+  include_directories : include_directories('.'))
```
Comment (jpakkane) [inline test cases/common/48 pkgconfig-gen/meson.build]: This is simply wrong. The current source and build directories are always guaranteed to be in the search path and our documentation explicitly says that you should not use `include_directories('.')`. This needs a different solution.

File: docs/markdown/Pkgconfig-module.md [hunk 7]
```diff
@@ -52,10 +52,20 @@ keyword arguments.
...
 - `description` is set to the project's name followed by the library's name.
 - `name` is set to the library's name.
+
+Since 0.49.0 uninstalled pkg-config files are generated as well. They are
+located in `<build dir>/pkgconfig-uninstalled/`. It is sometimes
+useful to build projects against libraries built by meson without having to
+install them into a prefix. In order to do so, just set
+`PKG_CONFIG_PATH=<builddir>/pkgconfig-uninstalled` before building your
+application. That will cause pkg-config to prefer those `-uninstalled.pc` files
+and find libraries and headers from the meson builddir.
```
Comment (jpakkane) [inline docs/markdown/Pkgconfig-module.md]: This should mention that uninstalled is provided on a best-effort basis and that there is no guarantee that it will work for any specific use case (or keep working if it currently does) because there are many ways that uninstalled libs can fail and we can't support all of them. It should instead recommend subprojects or staged install when possible.

File: docs/markdown/snippets/uninstalled-pkgconfig.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+## Uninstalled pkg-config files
+
+The `pkgconfig` module now generates uninstalled pc files as well. For any generated
+`foo.pc` file, an extra `foo-uninstalled.pc` file is placed into
+`<builddir>/pkgconfig-uninstalled`. They can be used to build applications against
+libraries built by meson without installing them, by pointing `PKG_CONFIG_PATH`
+to that directory.
```
Comment (jpakkane) [inline docs/markdown/snippets/uninstalled-pkgconfig.md]: Same caveats here.

File: mesonbuild/modules/pkgconfig.py (hunk 4 above)
Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: It's a good idea, so I gave it a try but the unit test didn't pass, I'm not sure why.

The generated `libfoo-uninstalled.pc` contains `Libs: /tmp/tmpuhj3659h/libfoo.so.1.0`. That file exists. But when I run the test app that links on that lib it says: `/tmp/tmpmvxgab95/test1: error while loading shared libraries: libfoo.so.1: cannot open shared object file: No such file or directory`.

Looks like we don't support having full path in pkg-config files. Probably we keep only the basename but don't add the rpath, or something like that?

File: mesonbuild/modules/pkgconfig.py (hunk 5 above)
Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: Right, renamed that function. I'll reply below about the include directories.

File: test cases/common/48 pkgconfig-gen/meson.build (hunk 6 above)
Comment (xclaesse) [reply test cases/common/48 pkgconfig-gen/meson.build]: Right, I changed to always add `-I` cflag to the library's subdir in both source and build trees implicitly.

File: docs/markdown/Pkgconfig-module.md (hunk 7 above)
Comment (xclaesse) [reply docs/markdown/Pkgconfig-module.md]: done.

File: docs/markdown/snippets/uninstalled-pkgconfig.md (hunk 8 above)
Comment (xclaesse) [reply docs/markdown/snippets/uninstalled-pkgconfig.md]: done

File: mesonbuild/modules/pkgconfig.py (hunk 4 above)
Comment (textshell) [reply mesonbuild/modules/pkgconfig.py]: I've just tried with my meson wrapper to make sure there are no regressions. But it still works.
One difference i see is that my wrapper generates 
```
Libs: /home/.../toolkit/_build/src/libtuiwidgets.so
```

So it uses the .so name going through the symlinks. But i can't really imagine why that should make a difference on linux.

**Comment by xclaesse** [comment]
Updated this PR.

I changed the directory for uninstalled pc files to `<builddir>/meson-uninstalled` to use the `meson-` prefix like we already have for `meson-logs` and `meson-private`, and also because we could maybe need to generate other 'uninstalled' helpers stuff some day. 

It now implicitly adds `-I<source_dir>/<library_subdir>` and `-I<builddir_dir>/<library_subdir>` in Cflags, because it's common to have headers in the same dir as source files, and we already have an implicit `include_directories('.')` at build time. That should also cover the case we have generated headers.

> That said i think we need to come up with a good solution for the include paths in the uninstalled variant. I believe we might have to some some restrictions on possible directory structures. I.e. if the includes are restructured on install i think we should not try to support that. But if a project wants to have a good uninstalled story and places all includes already in suitable directories we should support that.

It is really common to have headers in the same directory as source files, then install them all into the same location. That's what GStreamer does and I'm working on this PR specifically for that project. Of course that means that when using uninstalled pc files the app could include private headers that are not normally installed. But there is nothing we can do against that, I don't think it's a big deal.

File: docs/markdown/Pkgconfig-module.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (xclaesse) [reply docs/markdown/Pkgconfig-module.md]: Since there is no good solution, let's just keep things as they are. I don't think its a big deal in practice.

File: mesonbuild/modules/pkgconfig.py (hunk 4 above)
Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: It's the rpath system that doesn't work with versioned soname. See the difference in the command line ninja executes:

```
[7/10] ccache cc  -o test1 'test1@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed -Wl,--start-group /tmp/tmpogah5u7m/libfoo.so.1.0 -Wl,--end-group  
```
and
```
[7/10] ccache cc  -o test1 'test1@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed -Wl,--start-group /tmp/tmpytj5nrxu/libfoo.so -Wl,--end-group -Wl,-rpath,/tmp/tmpytj5nrxu -Wl,-rpath-link,/tmp/tmpytj5nrxu  
```

Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: I think it's a bug in rpaths_for_bundled_shared_libraries():
```
            if os.path.splitext(libpath)[1] not in ['.dll', '.lib', '.so', '.dylib']:
                continue
```

The extension of `libfoo.so.1` is `.1`...

Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: Voilà, added a commit to fix that bug.

File: mesonbuild/backend/backends.py [hunk 9]
```diff
@@ -380,8 +380,10 @@ def rpaths_for_bundled_shared_libraries(self, target, exclude_system=True):
                 # No point in adding system paths.
                 continue
             # Windows doesn't support rpaths, but we use this function to
-            # emulate rpaths by setting PATH, so also accept DLLs here
-            if os.path.splitext(libpath)[1] not in ['.dll', '.lib', '.so', '.dylib']:
+            # emulate rpaths by setting PATH, so also accept DLLs here.
+            # Also match soname in the form libfoo.so[.X[.Y[.Z]]]
+            if os.path.splitext(libpath)[1] not in ['.dll', '.lib', '.dylib'] and \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__ <sub>posted by Sider</sub>

[W605] invalid escape sequence '\.'

File: mesonbuild/modules/pkgconfig.py [hunk 10]
```diff
@@ -272,6 +283,11 @@ def generate_libs_flags(libs):
                 for l in libs:
                     if isinstance(l, str):
                         yield l
+                    elif uninstalled:
+                        path = state.backend.get_target_filename_abs(l)
+                        if not path in Lflags:
```
Comment (ignatenkobrain) [inline mesonbuild/modules/pkgconfig.py]: __[Flake8]__ <sub>posted by Sider</sub>

[E713] test for membership should be 'not in'

**Comment by xclaesse** [comment]
@jpakkane @textshell All review comments have been fixed.

File: mesonbuild/backend/backends.py [hunk 11]
```diff
@@ -380,8 +380,10 @@ def rpaths_for_bundled_shared_libraries(self, target, exclude_system=True):
...
+               not re.match(r'.*\.so(\.[0-9]+)?(\.[0-9]+)?(\.[0-9]+)?$', libpath):
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: but what if SONAME is libfoo-2.so?

Comment (xclaesse) [reply mesonbuild/backend/backends.py]: that regext catch <anything>.so<soname>

**Comment by xclaesse** [comment]
After some discussion on IRC, turns out I didn't understand @textshell's suggestion. What he is suggesting is to not add all include directories from all libraries, but instead add all source directories from `install_headers()`.

I think in practice that would yield of subset of the include directories, with potentially some private include directories excluded from the set of include paths which would be a good thing. However:
- We cannot really prevent apps using -uninstalled.pc to include private headers. It only makes a difference for projects where public and private headers are in different directories in the source tree.
- The `install_headers()` call could be after `pkg.generate()` call, so that means we'll have to postpone uninstalled pkg-config generation to the end of configure. That likely means more invasive and complicated implementation. We could also just assume `install_headers()` must be called before calling `pkg.generate()` otherwise -uninstalled.pc will be missing some include directories. In practice that's often the case, but that looks ugly to me.
- Since we cannot know which `install_headers()` call is for which library, we'll have to include them all into every -uninstalled.pc file. That means having unrelated paths, but probably not that bad.

In conclusion, I agree it's a better solution, but I'm not convinced it's worth the effort. My fear is doing more invasive changes won't be accepted by @jpakkane who is already not really fond on this.

File: run_unittests.py [hunk 12]
```diff
@@ -4746,6 +4746,21 @@ def test_pkgconfig_gen_deps(self):
...
         self.assertEqual(out, ['-llibmain2', '-llibinternal'])
 
+    def test_pkgconfig_uninstalled(self):
+        testdir = os.path.join(self.common_test_dir, '48 pkgconfig-gen')
+        self.init(testdir)
+        self.build()
+
+        os.environ['PKG_CONFIG_LIBDIR'] = os.path.join(self.builddir, 'meson-uninstalled')
+        if is_cygwin():
+                os.environ['PATH'] += os.pathsep + self.builddir
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__ over-indented ([see on Sider](https://sider.review/gh/repos/19784232/pulls/4436#issue-4869541))

<details>

| Rule |
| - |
| E117 |

</details>
<sub>posted by Sider</sub>

**Comment by xclaesse** [comment]
I rebased my branch, and reverted to the old version before the changes described in https://github.com/mesonbuild/meson/pull/4436#issuecomment-451508300. The reason is those changes makes the patch a lot more complicated and invasive (and merge conflicts) for little gains. I would like to get this merged first, then we can always improve implementation if issues are spotted.

One reason I'm revisiting this patch now is because a new use-case appeared: GStreamer has a repository with rust plugins, built with cargo. We want to build those plugins using cargo as a meson subproject by running cargo as a custom_target() with PKG_CONFIG_PATH pointing to meson-uninstalled directory. That way cargo will build plugins against the gstreamer meson is building. We have a working POC: https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/merge_requests/205

**Comment by xclaesse** [comment]
@textshell @jpakkane ping? I think this is ready to merge.

**Comment by xclaesse** [comment]
@textshell @jpakkane if nobody object, I'll just go ahead and merge this.

**Comment by textshell** [comment]
I'm sorry that i can't commit to a timely rereview of this. But from what i have seen nothing substantial changed since the time where i voiced my objections.

I know this is very frustrating. I'm sorry for that. But i don't think your last comment is in the least helpful, as you very well know that the contribution guidelines explicitly require an "ack" from @jpakkane for new features and not just silence after voicing at least major discomfort.

**Comment by jpakkane** [comment]
I have never particularly liked the uninstalled setup, because it is a maintenance burden. Basically what we are doing here is working around the way Cargo has steadfastly refused to integrate with any other workflow than their own.

**Comment by xclaesse** [comment]
> I'm sorry that i can't commit to a timely rereview of this. But from what i have seen nothing substantial changed since the time where i voiced my objections.

While I agree your suggestion is marginally better, the implementation is significantly harder and is a lot more intrusive. I did implement it (see history of this PR) but nobody reviewed it, and it conflicted badly. I decided to revert back to original idea to keep it simple.

> I know this is very frustrating. I'm sorry for that. But i don't think your last comment is in the least helpful, as you very well know that the contribution guidelines explicitly require an "ack" from @jpakkane for new features and not just silence after voicing at least major discomfort.

Side note: As the silence on this PR shows, I think @jpakkane as way too much on his plate already, we should learn to delegate more in the Meson project.

> I have never particularly liked the uninstalled setup, because it is a maintenance burden.

That's exactly the reason why I reverted this PR to the much less invasive implementation. This patch is pretty well self contained, it does not interfere with the rest of the Meson code base. As for the maintenance burden, I'm totally happy to take responsibility of the pkgconfig module since I perfectly understand you already have too much to work on. If someone report an issue against uninstalled.pc files, just leave it to me ;-)

> Basically what we are doing here is working around the way Cargo has steadfastly refused to integrate with any other workflow than their own.

This cargo stuff is just one new type of use-case where I found uninstalled pc files could be useful. But the main use-case remains a developer tool. This has been blocking GStreamer to switch to pkgconfig generator for over a year now because some developers rely on uninstalled pc files in their workflow. https://gitlab.freedesktop.org/gstreamer/gstreamer/merge_requests/4.

**Comment by xclaesse** [comment]
Ping again, it's sad to miss again the release deadline here. I've got patches waiting on gstreamer blocked by this, and also rust integration that would profit a lot of it. Given how long this has been waiting, I'm really tempted to just merge it and commit myself in fixing any potential regression.

Btw, I'm also offering to become maintainer of the pkgconfig module, given that I wrote a big part of its logic, in the effort to delegate more responsibilities in the meson project.

**Comment by sdroege** [comment]
This works great for me, there's just one problem with other software parsing the output of pkg-config: currently you set the whole path to the shared library in the `Libs` field. Various other software out there seems to assume that it will contain the library as `-lfoo` instead there. By having `-L/path/to/uninstalled/so -lfoo` in there instead this would work fine again.

Was there a specific reason to not do that?

**Comment by xclaesse** [comment]
By "various other software" you mean cargo at least l, right? I'll get a look at that, I wrote it a long time ago.

**Comment by sdroege** [comment]
The Rust `pkg-config` crate parses the output and converts it to something `rustc` understands (and so that it works for MSVC and other compilers). I saw similar code parsing the `-L` and `-l` flags in other places too though.

**Comment by nirbheek** [comment]
I would also like to see this feature in Meson, but I can't officially give a +1 because I haven't really been contributing to Meson lately 🙂

**Comment by xclaesse** [comment]
Ok, changed the implementation to use -L -l pairs instead, not sure why I didn't do that from the beginning, all the code was already there.

I also changed to have ${prefix} set to build directory, and ${srcdir} set to source directory, and made all include paths using them instead of repeating long full paths everywhere. That makes the pc file more readable.

**Comment by sdroege** [comment]
Thanks, looks good to me :) Just one minor issue

```
Libs: -L${prefix}/src  -ldav1d
```

There are 2 spaces between the `-L` and `-l`

**Comment by xclaesse** [comment]
> There are 2 spaces between the `-L` and `-l`

That's unrelated, it's like that in regular pc files too. But fixed it now :)

Also updated version to 0.54 since we missed the 0.53 release.

**Comment by gdesmott** [comment]
We'd need this feature in order to be able to build Rust GStreamer plugins on our Windows ci. Any chance to have it merged before the next meson release?

**Comment by sdroege** [comment]
Any updates here? We'd also need this for `gst-uninstalled` to work in a useful way (especially on Windows) where other meson-based projects (like GLib) are built as part of the whole thing and are not available via `pkg-config` unless everything is installed.
